keps/sig-api-machinery/5000-api-linting-crd-schema-tooling/README.md

```diff
@@ -0,0 +1,959 @@
+- **commentstart**: checks that all comments in the API types starts with the serialized form of the type they are commenting on.
+- **integers**: checks for usage of unsupported integer types.
+- **jsontags**: checks that all fields in API types have a `json` tag and that the `json` tags are correctly formatted.
+- **maxlength**: checks that string and array fields in the API are bounded by a maximum length.
+- **nobools**: checks fields in the API types do not contain a `bool` type.
+- **nophase**: checks that fields in the API types don't contain a `Phase` or any field where `Phase` is a substring.
+- **optionalorrequired**: checks that all fields in an API type are marked as either optional or required.
+- **requiredfields**: checks that all fields in an API type marked as required follow the convention of not being pointers and not having `omitempty` in their JSON tag.
+- **statussubresource**: checks that the status subresource is correctly configured correctly for API types that back a CRD.
+
+Most implemented linters have configuration options, and all linters can be enabled/disabled as desired in a golangci-lint configuration file.
+
+### CRD Schema Change Validation
+
+There is some previous work done in both https://github.com/everettraven/crd-diff and https://github.com/openshift/crd-schema-checker. Each project implements schema checks in different ways and has different user workflows.
+
+It is proposed that a new repository is created that merges select bits and pieces from each project into a single project.
+
+Currently, it is proposed that:
+- The CLI UX from https://github.com/everettraven/crd-diff is used. It currently allows sourcing both the old and new CRDs from various locations, allowing for use case flexibility.
+- The reporting logic from https://github.com/openshift/crd-schema-checker is used. It enforces that all checks include a description of why the check matters and allows for communicating errors, warnings, and other important information found during the comparison process.
+- The CRD manifest schema walking logic from https://github.com/openshift/crd-schema-checker is used. https://github.com/everettraven/crd-diff imports this logic.
+- The validation/validator pattern from https://github.com/everettraven/crd-diff. It clearly distinguishes between individual property validations and broader scoped CRD validations. Additionally, it has a validation pattern in place that checks for general CRD changes, compatibility of version-to-version, and compatibility of served versions.
+
+https://github.com/everettraven/crd-diff currently has some configuration logic in place, but it is highly repetitive and not a great user experience. This needs to be evaluated further and will likely result in a distinct configuration process for the new project.
+
+Additionally, the following validations from both projects are proposed to be included in the move to the common project:
```

Comment:
We probably want to list the checks that exist in these projects, that we don't want to carry over, and why

I know OLM had some pre-upgrade checks that were implemented in a different way, and quite a few of those don't exist in `crd-schema-checker`, are they all captured here?

Reply:
Should be updated in https://github.com/kubernetes/enhancements/pull/5102/commits/ebfd4d1bdb572594ed4a50d08e6b9a43c783f476